src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,89 @@ public boolean wasRetried() {
     this.m_wasRetried = wasRetried;
   }
+
+  public List<ITestNGMethod> getSkipCausedBy() {
+    if (this.m_status != SKIP || skipAnalysed) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    skipAnalysed = true;
+    //check if there were any config failures
+    Set<ITestResult> skippedConfigs = m_context.getFailedConfigurations().getAllResults();
+    for (ITestResult skippedConfig : skippedConfigs) {
+      if (isGlobalFailure(skippedConfig) || isRelated(skippedConfig)) {
+        //If there's a failure in @BeforeTest/@BeforeSuite/@BeforeClass
+        //then the reason is most often just one method.
+        skippedDueTo.add(skippedConfig.getMethod());
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      if (belongToSameGroup(skippedConfig)) {
+        //If its @BeforeGroups then there's a chance that there could be more than one
+        //method. So lets add everything.
+        skippedDueTo.add(skippedConfig.getMethod());
+      }
+    }
+    if (!skippedDueTo.isEmpty()) {
+      //If we found atleast one skipped due to reason, then its time to return back.
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    //Looks like we didnt have any configuration failures. So some upstream method perhaps failed.
+    List<String> upstreamMethods = Arrays.asList(m_method.getMethodsDependedUpon());
+    if (upstreamMethods.isEmpty()) {
+      //Maybe group dependencies exist ?
+      List<String> upstreamGroups = Arrays.asList(m_method.getGroupsDependedUpon());
+      if (upstreamGroups.isEmpty()) {
+        return Collections.emptyList();
+      }
+      List<ITestResult> allFailures = Lists.merge(m_context.getFailedTests().getAllResults(),
+          m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+      skippedDueTo = allFailures.stream()
+          .map(ITestResult::getMethod)
+          .filter(method -> {
+            List<String> currentMethodGroups = Arrays.asList(method.getGroups());
+            List<String> interection = Lists.intersection(upstreamGroups, currentMethodGroups);
+            return !interection.isEmpty();
+          }).collect(Collectors.toList());
+
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+
+    //So we have dependsOnMethod failures
+    List<ITestResult> allfailures = Lists.merge(m_context.getFailedTests().getAllResults(),
+        m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+    skippedDueTo = allfailures.stream()
+        .map(ITestResult::getMethod)
+        .filter(method -> upstreamMethods.contains(method.getQualifiedName()))
+        .collect(Collectors.toList());
+    return Collections.unmodifiableList(skippedDueTo);
+  }
+
+  private static boolean isGlobalFailure(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    return m.isBeforeTestConfiguration() || m.isBeforeSuiteConfiguration();
+  }
+
+  private boolean isRelated(ITestResult result) {

```

Comment:
I think the signature should be `private static boolean isRelated(ITestResult result, Object instance)` because another option could be `boolean ITestResult#isRelated(Object instance)`

Reply:
I am not quite sure I understand this. Here `isRelated()` is basically comparing just two `ITestResult` objects. That is why its an instance method. If it were to be static, then I would need to basically have it take in two `ITestResult` objects.

Reply:
Ok